File: solutions/java/LongestIncreasingSubarray.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Random;

public class LongestIncreasingSubarray {
  // @include
  // Represent subarray by starting and ending indices, inclusive.
  private static class Subarray {
    public Integer start;
    public Integer end;

    public Subarray(Integer start, Integer end) {
      this.start = start;
      this.end = end;
    }
  }

  public static Subarray findLongestIncreasingSubarray(List<Integer> A) {
    int maxLength = 1;
    Subarray result = new Subarray(0, 0);
    int i = 0;
    while (i < A.size() - maxLength) {
      // Backward check and skip if A[j - 1] >= A[j].
      boolean isSkippable = false;
      for (int j = i + maxLength; j > i; --j) {
        if (A.get(j - 1) >= A.get(j)) {
          i = j;
          isSkippable = true;
          break;
        }
      }

      // Forward check if it is not skippable.
      if (!isSkippable) {
        i += maxLength;
        while (i < A.size() && A.get(i - 1) < A.get(i)) {
          ++i;
          ++maxLength;
        }
        result = new Subarray(i - maxLength, i - 1);
      }
    }
    return result;
  }
  // @exclude

  private static void simpleTest() {
    Subarray result = findLongestIncreasingSubarray(Arrays.asList(-1, -1));
    assert(result.start == 0 && result.end == 0);
    result = findLongestIncreasingSubarray(Arrays.asList(1, 2));
    assert(result.start == 0 && result.end == 1);
  }

  public static void main(String[] args) {
    simpleTest();
    Random gen = new Random();

    for (int times = 0; times < 1000; ++times) {
      List<Integer> A = new ArrayList<>();
      if (args.length > 2) {
        for (int i = 1; i < args.length; ++i) {
          A.add(Integer.parseInt(args[i]));
        }
      } else {
        int n;
        if (args.length == 1) {
          n = Integer.parseInt(args[0]);
        } else {
          n = gen.nextInt(10000) + 1;
        }
        for (int i = 0; i < n; ++i) {
          A.add((gen.nextBoolean() ? -1 : 1) * gen.nextInt(n));
        }
      }
      Subarray result = findLongestIncreasingSubarray(A);
      System.out.println(result.start + " " + result.end);
      int len = 1;
      for (int i = 1; i < A.size(); ++i) {
        if (A.get(i) > A.get(i - 1)) {
          ++len;
        } else {
          len = 1;
        }
        assert(len <= result.end - result.start + 1);
      }
    }
  }
}

File: solutions/python/longest_increasing_subarray.py
import collections
import sys
import random

# @include
Subarray = collections.namedtuple('Subarray', ('start', 'end'))


def find_longest_increasing_subarray(A):
    result = Subarray(0, 0)
    i, max_length = 0, 1
    while i < len(A) - max_length:
        # Backward check and skip if A[j] >= A[j + 1].
        for j in range(i + max_length, i, -1):
            if A[j - 1] >= A[j]:
                i = j
                break
        else:  # Forward check if it is not skippable (the loop ended normally)
            i += max_length
            while i < len(A) and A[i - 1] < A[i]:
                i, max_length = i + 1, max_length + 1
            result = Subarray(i - max_length, i - 1)
    return result
# @exclude


def simple_test():
    result = find_longest_increasing_subarray([-1, -1])
    assert result.start == 0 and result.end == 0
    result = find_longest_increasing_subarray([1, 2])
    assert result.start == 0 and result.end == 1


def main():
    simple_test()
    for _ in range(1000):
        if len(sys.argv) > 2:
            A = list(map(int, sys.argv[1:]))
        else:
            n = int(sys.argv[1]) if len(sys.argv) == 2 else random.randint(
                1, 1000000)
            A = [random.randint(-(n - 1), n - 1) for _ in range(n)]
        result = find_longest_increasing_subarray(A)
        print(*result)
        max_len = 1
        cur_len = 1
        for i in range(1, len(A)):
            if A[i] > A[i - 1]:
                cur_len += 1
                max_len = max(max_len, cur_len)
            else:
                cur_len = 1
        assert max_len == result.end - result.start + 1


if __name__ == '__main__':
    main()
